src/dropck-eyepatch.md

```diff
@@ -0,0 +1,434 @@
+struct Inspector<'a, T: 'a> { data: &'a T }
+
+impl<'a, T> Drop for Inspector<'a, T> {
+    fn drop(&mut self) {
+        println!("I was only one day from retirement!");
+    }
+}
+
+fn main() {
+    let (data, inspector);
+    data = Box::new(0u8);
+    inspector = Inspector { data: &*data };
+}
+```
+
+This Inspector is perfectly safe, because it doesn't actually access its
+generic data in its destructor. Sadly, the code still doesn't compile:
+
+```text
+error: `*data` does not live long enough
+  --> src/main.rs:13:1
+   |
+12 |     inspector = Inspector { data: &*data };
+   |                                    ----- borrow occurs here
+13 | }
+   | ^ `*data` dropped here while still borrowed
+   |
+   = note: values in a scope are dropped in the opposite order they are created
+```
+
+This is because as far as Rust is concerned you *could have* accessed it, and
+Rust refuses to inspect your drop implementation to be sure.
+
+With [the eyepatch RFC][eyepatch], we can *partially blind* dropck, by hiding one of our
+generic parameters from it. (...by covering it with a patch. Get it? ...Eyepatch?)
+
+The patch we apply is as follows:
+
+```rust
+#![feature(generic_param_attrs, dropck_eyepatch)]
+
+struct Inspector<'a, T: 'a> { data: &'a T }
+
+// Changes here:
+unsafe impl<#[may_dangle] 'a, T> Drop for Inspector<'a, T> {
+    fn drop(&mut self) {
+        println!("I was only one day from retirement!");
+    }
+}
+
+fn main() {
+    let (data, inspector);
+    data = Box::new(0u8);
+    inspector = Inspector { data: &*data };
+}
+```
+
+...and it compiles and runs!
+
+There are two changes:
+
+* We add `#[may_dangle]` to one of our type parameters
+* We add `unsafe` to the impl block (which is required by may_dangle to emphasize the risks)
+
+Note also that `#[may_dangle]` requires both the `generic_param_attrs`, and the
+`dropck_eyepatch` features.
+
+The may_dangle attribute tells the dropck to ignore `'a` in its analysis. Since this was
+the only reason the Inspector was considered unsound (T is just a u8), our code compiles.
+
+`#[may_dangle]` may be applied to any type parameter. For instance, if we change
+`data` to just `T` (so `T = &'a u8`), then we need to blind dropck from `T`:
+
+```rust
+#![feature(generic_param_attrs, dropck_eyepatch)]
+
+struct Inspector<T> { data: T }
+
+// Changes here:
+unsafe impl<#[may_dangle] T> Drop for Inspector<T> {
+    fn drop(&mut self) {
+        println!("I was only one day from retirement!");
+    }
+}
+
+fn main() {
+    let (data, inspector);
+    data = Box::new(0u8);
+    inspector = Inspector { data: &*data };
+}
+```
+
+
+
+
+# When The Eyepatch is (Un)Sound
+
+The general rule for when it's safe to apply the dropck eyepatch to a type parameter
+`T` is that the destructor must only do things to values of type `T` that could be
+done with *all* types. Basically: we can move (or copy) the values around, take
+references to them, get their size/align, and drop them. Just to be clear
```

Comment:
This is discussed in the next section, but I should emphasize this isn't quite true here